### PR TITLE
Update to 1.4

### PR DESCRIPTION
**Pull request by SokyranTheDragon**

### CODE REVIEW

File: Source/Client/AsyncTime/AsyncTimeComp.cs [hunk 1]
```diff
@@ -173,6 +173,8 @@ public void UpdateManagers()
 
             map.powerNetManager.UpdatePowerNetsAndConnections_First();
             map.glowGrid.GlowGridUpdate_First();
+
+            map.postTickVisuals.ProcessPostTickVisuals();
```
Comment (SokyranTheDragon) [inline Source/Client/AsyncTime/AsyncTimeComp.cs:177]: Stuff like calculating pawn rotation, as well as other stuff. It was moved from pawn ticking to map update.

File: Source/Client/Debug/DebugTools.cs [hunk 2]
```diff
@@ -1,409 +1,409 @@
...
-        }
-
-        public static int Hash(this FloatMenuOption opt)
-        {
-            return Gen.HashCombineInt(GenText.StableStringHash(opt.action.Method.MethodDesc()), GenText.StableStringHash(opt.labelInt));
-        }
-    }
-
-}
+// using System;
```
Comment (SokyranTheDragon) [inline Source/Client/Debug/DebugTools.cs:1]: Dev tools and options were changed quite a bit in 1.4, and I've been unable to fix them.

File: Source/Client/Patches/Determinism.cs [hunk 3]
```diff
@@ -123,7 +120,7 @@ static void Postfix(ref FloodUnfogResult __result)
         }
     }
 
-    [HarmonyPatch(typeof(Pawn_DrawTracker), nameof(Pawn_DrawTracker.DrawTrackerTick))]
+    [HarmonyPatch(typeof(Pawn), nameof(Pawn.ProcessPostTickVisuals))]
```
Comment (SokyranTheDragon) [inline Source/Client/Patches/Determinism.cs:123]: The code this was patching was moved from Pawn_DrawTracker to Pawn class

File: Source/Client/Patches/Determinism.cs [hunk 4]
```diff
@@ -136,8 +133,8 @@ static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstruction> inst
 
                 if (inst.operand == CellRectContains)
                 {
-                    yield return new CodeInstruction(OpCodes.Ldc_I4_1);
-                    yield return new CodeInstruction(OpCodes.Or);
+                    yield return new CodeInstruction(OpCodes.Ldc_I4_0);
+                    yield return new CodeInstruction(OpCodes.And);
```
Comment (SokyranTheDragon) [inline Source/Client/Patches/Determinism.cs:137]: The condition was inverted in 1.4, so we invert our condition as well. I'm rather sure this is correct, however I'd appreciate if someone was able to double check it.

File: Source/Client/Patches/Determinism.cs [hunk 5]
```diff
@@ -172,36 +169,36 @@ static bool ShouldShuffle()
...
-
-            return list;
-        }
-
-        static int Value(Bill bill, Pawn pawn)
-        {
-            return FloatMenuMakerMap.makingFor == pawn ? bill.lastIngredientSearchFailTicks : 0;
-        }
-    }
+    // [HarmonyPatch(typeof(WorkGiver_DoBill), nameof(WorkGiver_DoBill.StartOrResumeBillJob))]
```
Comment (SokyranTheDragon) [inline Source/Client/Patches/Determinism.cs:172]: The way this works in 1.4 was changed a bit. It no longer stores the last fail tick, instead storing the next tick to do a check. It never sets the value to 0 like it was done in 1.3

File: Source/Client/Patches/StylingStation.cs [hunk 6]
```diff
@@ -198,7 +198,6 @@ internal static void StylingDialog_Accept(StylingData data)
                 else
                 {
                     pawn.style.Notify_StyleItemChanged();
-                    dialog.MakeHairFilth();
```
Comment (SokyranTheDragon) [inline Source/Client/Patches/StylingStation.cs:201]: Seems like this method was removed in 1.4, and there was no replacement for it in the dialog itself

File: Source/Client/Patches/TickPatch.cs [hunk 7]
```diff
@@ -186,6 +187,7 @@ private static void TickTickable(ITickable tickable)
         {
             while (tickable.RealTimeToTickThrough >= 0)
             {
+                Find.TickManager.ticksThisFrame++;
```
Comment (SokyranTheDragon) [inline Source/Client/Patches/TickPatch.cs:190]: Required for actually calculating some graphical stuff, like pawn rotation. Those are only called if the amount of ticks was not 0.

File: Source/Client/Persistent/CaravanFormingProxy.cs [hunk 8]
```diff
@@ -43,6 +43,22 @@ public override void DoWindowContents(Rect inRect)
                 drawing = null;
             }
         }
+
+        // This method was removed in 1.4
```
Comment (SokyranTheDragon) [inline Source/Client/Persistent/CaravanFormingProxy.cs:47]: Just to clarify - it was removed as a method, instead being called manually

File: Source/Client/Syncing/Dict/SyncDictRimWorld.cs [hunk 9]
```diff
@@ -266,13 +266,9 @@ public static class SyncDictRimWorld
...
                             sync.Write(ability);
                         }
-                        else if (verb.DirectOwner is CompEquippable compEquippable) {
-                            sync.Write(VerbOwnerType.CompEquippable);
-                            sync.Write(compEquippable);
-                        }
-                        else if (verb.DirectOwner is CompReloadable compReloadable) {
-                            sync.Write(VerbOwnerType.CompReloadable);
-                            sync.Write(compReloadable);
+                        else if (verb.DirectOwner is ThingComp thingComp) {
```
Comment (SokyranTheDragon) [inline Source/Client/Syncing/Dict/SyncDictRimWorld.cs:269]: I felt it was a bit redundant to check for an exact ThingComp, instead of syncing it as ThingComp itself. This should also improve compatibility with mods that use custom comps that hold verbs.

File: Source/Client/Syncing/Game/SyncDelegates.cs [hunk 10]
```diff
@@ -19,18 +16,32 @@ public static void Init()
...
             SyncDelegate.Lambda(typeof(FloatMenuMakerMap), nameof(FloatMenuMakerMap.AddDraftedOrders), 3).CancelIfAnyFieldNull().SetContext(mouseKeyContext);       // Drafted carry to bed
             SyncDelegate.Lambda(typeof(FloatMenuMakerMap), nameof(FloatMenuMakerMap.AddDraftedOrders), 4).CancelIfAnyFieldNull().SetContext(mouseKeyContext);       // Drafted carry to bed (arrest)
             SyncDelegate.Lambda(typeof(FloatMenuMakerMap), nameof(FloatMenuMakerMap.AddDraftedOrders), 5).CancelIfAnyFieldNull().SetContext(mouseKeyContext);       // Drafted carry to transport shuttle
             SyncDelegate.Lambda(typeof(FloatMenuMakerMap), nameof(FloatMenuMakerMap.AddDraftedOrders), 6).CancelIfAnyFieldNull().SetContext(mouseKeyContext);       // Drafted carry to cryptosleep casket
 
+            var extractRelicToArea = SyncDelegate.Lambda(typeof(FloatMenuMakerMap), nameof(FloatMenuMakerMap.AddHumanlikeOrders), 18);                   // Extract relic to area
+            extractRelicToArea.CancelIfAnyFieldNull().SetContext(mouseKeyContext);
+            extractRelicToArea.TransformField("job", Serializer.New(
```
Comment (SokyranTheDragon) [inline Source/Client/Syncing/Game/SyncDelegates.cs:34]: It's not possible to chain CancelIfAnyFieldNull and TransformField.

File: Source/Client/Syncing/Game/SyncMethods.cs [hunk 11]
```diff
@@ -159,7 +159,7 @@ public static void Init()
 
             SyncMethod.Register(typeof(CompAnimalPenMarker), nameof(CompAnimalPenMarker.RemoveForceDisplayedAnimal));
             SyncMethod.Register(typeof(CompAnimalPenMarker), nameof(CompAnimalPenMarker.AddForceDisplayedAnimal));
-            SyncMethod.Register(typeof(CompAnimalPenMarker), nameof(CompAnimalPenMarker.DesignatePlantsToCut));
+            SyncMethod.Register(typeof(CompAutoCut), nameof(CompAnimalPenMarker.DesignatePlantsToCut));
```
Comment (SokyranTheDragon) [inline Source/Client/Syncing/Game/SyncMethods.cs:162]: Method was moved from CompAnimalPenMarker to CompAutoCut

File: Source/Client/Windows/ModCompatWindow.cs [hunk 12]
```diff
@@ -391,7 +391,7 @@ enum SortDirection
     }
 
     [HotSwappable]
-    [HarmonyPatch(typeof(Page_ModsConfig), nameof(Page_ModsConfig.DoWindowContents))]
+    // [HarmonyPatch(typeof(Page_ModsConfig), nameof(Page_ModsConfig.DoWindowContents))]
```
Comment (SokyranTheDragon) [inline Source/Client/Windows/ModCompatWindow.cs:394]: The mod list was changed, so it'll need changes to this patch.

**Review by notfood — APPROVED** [review]

File: Source/Client/Syncing/Game/SyncMethods.cs (hunk 11 above)
Comment (Zetrith) [reply Source/Client/Syncing/Game/SyncMethods.cs:162]: The method should also be CompAutoCut.DesignatePlantsToCut

File: Source/Client/Persistent/CaravanFormingProxy.cs (hunk 8 above)
Comment (Zetrith) [reply Source/Client/Persistent/CaravanFormingProxy.cs:47]: Isn't it now Notify_TransferablesChanged?

File: Source/Client/Patches/Determinism.cs (hunk 4 above)
Comment (Zetrith) [reply Source/Client/Patches/Determinism.cs:137]: I don't think it was inverted

File: Source/Client/MultiplayerData.cs [hunk 13]
```diff
@@ -68,7 +64,7 @@ public static ModContentPack DummyContentPack(ModMetaData mod)
             return new ModContentPack(mod.RootDir, mod.PackageId, mod.PackageIdPlayerFacing, 0, mod.Name, mod.Official);
         }
 
-        public static UniqueList<Texture2D> icons = new();
+        public static UniqueList<Texture> icons = new();
```
Comment (Zetrith) [inline Source/Client/MultiplayerData.cs:67]: Why this change?

Comment (SokyranTheDragon) [reply Source/Client/MultiplayerData.cs:67]: Type of `Command.icon` was changed in 1.4 to `Texture`, so it's either this, or alternatively we'd have to cast it to `Texture2D`. And since it's no longer guaranteed to be of the type `Texture2D` it could potentially be of a different type - however unlikely this may be.

File: Source/Client/Patches/Determinism.cs (hunk 4 above)
Comment (SokyranTheDragon) [reply Source/Client/Patches/Determinism.cs:137]: As I said, I may be wrong. However, as far as I understand it:

In 1.3 it was:
```csharp
if (Current.ProgramState == ProgramState.Playing && !Find.CameraDriver.CurrentViewRect.ExpandedBy(3).Contains(this.pawn.Position))
// Return early from DrawTrackerTick
```

However, in 1.4 it now is:
```csharp
if (Current.ProgramState != ProgramState.Playing || viewRect.Contains(base.Position))
// Call ProcessPostTickVisuals (replacement for DrawTrackerTick)
```

So in 1.3, our patch would end up making the condition always false (since it was negated), which always ended up in returning early:
```csharp
if (Current.ProgramState == ProgramState.Playing && !(Find.CameraDriver.CurrentViewRect.ExpandedBy(3).Contains(this.pawn.Position) | true))
```

In 1.4, the same patch would end up always continuing with the call to the method:
```csharp
if (Current.ProgramState != ProgramState.Playing || (viewRect.Contains(base.Position) | true))
```

File: Source/Client/Persistent/CaravanFormingProxy.cs (hunk 8 above)
Comment (SokyranTheDragon) [reply Source/Client/Persistent/CaravanFormingProxy.cs:47]: Ah, good catch. I've missed it. Let me fix this real quick.

File: Source/Client/Syncing/Game/SyncMethods.cs (hunk 11 above)
Comment (SokyranTheDragon) [reply Source/Client/Syncing/Game/SyncMethods.cs:162]: Good catch, I forgot to change it.

File: Source/Client/Patches/TickPatch.cs (hunk 7 above)
Comment (Zetrith) [reply Source/Client/Patches/TickPatch.cs:190]: Find.TickManager.ticksThisFrame should be set to 1 before every map.postTickVisuals.ProcessPostTickVisuals call.

File: Source/Client/AsyncTime/AsyncTimeComp.cs (hunk 1 above)
Comment (Zetrith) [inline Source/Client/AsyncTime/AsyncTimeComp.cs:177]: This should be called in Tick().
It being in UpdateManagers makes it confusing as UpdateManagers runs in Tick and ExecuteCmd (see the code comment) but ProcessPostTickVisuals checks for Find.TickManager.TicksThisFrame not being zero anyway.

File: Source/Client/Patches/TickPatch.cs (hunk 7 above)
Comment (SokyranTheDragon) [reply Source/Client/Patches/TickPatch.cs:190]: That's true. I tried to keep it close to how it's handled by the game in case some mods would rely on that value. However, in vanilla game it's only used by ProcessPostTickVisuals, so it would make sense to use it there. I assume we should set it to 0 right after as well, right?

File: Source/Client/AsyncTime/AsyncTimeComp.cs (hunk 1 above)
Comment (SokyranTheDragon) [reply Source/Client/AsyncTime/AsyncTimeComp.cs:177]: Hmm... that's true. I put it into `UpdateManagers` as my first thought was that it would fit, since this is where we call other a handful of methods from `MapUpdate`.

File: Source/Client/Patches/Determinism.cs (hunk 4 above)
Comment (Zetrith) [reply Source/Client/Patches/Determinism.cs:137]: It's a transpiler so you should look at the IL.

Also, I tested your code in-game and pawns, f.e., don't wiggle when downed.

With your C#,
`if (Current.ProgramState == ProgramState.Playing && !(Find.CameraDriver.CurrentViewRect.ExpandedBy(3).Contains(this.pawn.Position) | true))`
\=\=
`if (Current.ProgramState == ProgramState.Playing && false)`
\=\=
it never returns early
\=\=
it always runs

This patch is there so the pawn's DrawPos is always calculated (and not dependent on the camera position) as it affects things like bullet positions.

Comment (SokyranTheDragon) [reply Source/Client/Patches/Determinism.cs:137]: Oh my, yeah. That's a really big mistake on my part. Pushing a fix now.

File: Source/Client/AsyncTime/AsyncTimeComp.cs (hunk 1 above)
Comment (SokyranTheDragon) [reply Source/Client/AsyncTime/AsyncTimeComp.cs]: Fixed

File: Source/Client/Patches/TickPatch.cs (hunk 7 above)
Comment (SokyranTheDragon) [reply Source/Client/Patches/TickPatch.cs]: Done

File: Source/Client/MultiplayerStatic.cs [hunk 14]
```diff
@@ -292,14 +292,17 @@ void LogError(string str)
                 var randomBoltMesh = typeof(LightningBoltMeshPool).GetProperty(nameof(LightningBoltMeshPool.RandomBoltMesh)).GetGetMethod();
                 var drawTrackerCtor = typeof(Pawn_DrawTracker).GetConstructor(new[] { typeof(Pawn) });
                 var randomHair = typeof(PawnStyleItemChooser).GetMethod(nameof(PawnStyleItemChooser.RandomHairFor));
+                var cannotAssignReason = typeof(Dialog_BeginRitual).GetMethod(nameof(Dialog_BeginRitual.CannotAssignReason), BindingFlags.NonPublic | BindingFlags.Instance);
+                var canEverSpectate = typeof(RitualRoleAssignments).GetMethod(nameof(RitualRoleAssignments.CanEverSpectate));
 
                 var effectMethods = new MethodBase[] { subSustainerStart, sampleCtor, subSoundPlay, effecterTick, effecterTrigger, effecterCleanup, randomBoltMesh, drawTrackerCtor, randomHair };
                 var moteMethods = typeof(MoteMaker).GetMethods(BindingFlags.Static | BindingFlags.Public)
                     .Where(m => m.Name != "MakeBombardmentMote"); // Special case, just calls MakeBombardmentMote_NewTmp, prevents Hugslib complains
```
Comment (SokyranTheDragon) [inline Source/Client/MultiplayerStatic.cs:300]: Is this still needed? Seems like `MakeBombardmentMote_NewTmp` is no longer a thing in 1.4. Or in 1.3. However, `MakeBombardmentMote` doesn't call RNG, so (from what I understand) it should be deterministic already.

Speaking of which, isn't that also the case for several MoteMaker/FleckMaker methods, which either have no RNG, or only call other methods that may have it patched.

File: Source/Client/Persistent/Rituals.cs [hunk 15]
```diff
@@ -238,4 +234,30 @@ static bool Prefix(Window window)
         }
     }
 
+    // We do not sync Ability.QueueCastingJob - instead syncing the delegate method queueing the job.
+    // We do this to avoid confirmation dialogs from some of the abilities, like neuroquake and a few ones added in Biotech.
+    // This causes an issue with abilities like throne speech or leader speech - the ritual dialog itself is treated as the confirmation dialog.
+    // This patch exists to catch abilities that would create ritual "confirmation" dialogs, and instead syncs the call to QueueCastingJob instead.
+    // There is no need to do this for global abilities as there are none that create rituals, and (at least up to 1.4) there are no other abilities that create non-confirmation dialogs
+    [HarmonyPatch(typeof(Ability), nameof(Ability.QueueCastingJob), typeof(LocalTargetInfo), typeof(LocalTargetInfo))]
+    static class RedirectRitualDialogCast
```
Comment (SokyranTheDragon) [inline Source/Client/Persistent/Rituals.cs:243]: The comments should explain the reason behind this patch.

I couldn't really think of another way of handling this, at least with our current tool. I'm however open to suggestions on handling this in alternative ways.

One idea I've had would be if we had an option to determine if a method should be synced or not - like by slightly changing `SyncTemplates.General` to continue the method if `ISyncCall.DoSync` returned false. This would also require adding some extra code to `SyncMethod`. I've decided not to pursue this, as my past attempt at doing this ended up... not all that good, I'm not gonna lie. Anyway, anyone has any thoughts about this potential approach?

File: Source/Client/Persistent/TradingUI.cs [hunk 16]
```diff
@@ -223,20 +223,24 @@ private void BeforeCache()
             newTradeables.Clear();
         }
 
-        public static IEnumerable<Tradeable> AllTradeables()
+        public static List<Tradeable> AllTradeables()
```
Comment (SokyranTheDragon) [inline Source/Client/Persistent/TradingUI.cs:226]: I feel like I should explain the reason for this change - the `CacheTradeables` method (where we insert our `AllTradeables`) is now using `GenCollection.FirstOrDefault(this List<T> list)`, instead of calling `Enumerable.FirstOrDefault(this IEnumerable<TSource> source)` like it used to in 1.3 and earlier.

Since the type we've given to that method was not what it was expecting, this ended up crashing the game.

File: Source/Client/Syncing/Game/SyncDelegates.cs (hunk 10 above)
Comment (SokyranTheDragon) [reply Source/Client/Syncing/Game/SyncDelegates.cs]: Doesn't matter anymore, as I've completely dropped this sync and fixed in a different way

File: Source/Client/Syncing/Dict/SyncDictRimWorld.cs (hunk 9 above)
Comment (SokyranTheDragon) [reply Source/Client/Syncing/Dict/SyncDictRimWorld.cs:269]: Actually, thinking about this... Why not sync `Thing` instead of `Pawn`? Should improve compatibility with mods, and it doesn't really make much of a difference for us.

File: Source/Client/MultiplayerData.cs [hunk 17]
```diff
@@ -125,7 +121,6 @@ internal static void CollectDefInfos()
             dict["MapComponent"] = GetDefInfo(ImplSerialization.mapCompTypes, TypeHash);
 
             dict["PawnBio"] = GetDefInfo(SolidBioDatabase.allBios, b => b.name.GetHashCode());
-            dict["Backstory"] = GetDefInfo(BackstoryDatabase.allBackstories.Keys, b => b.GetHashCode());
```
Comment (SokyranTheDragon) [inline Source/Client/MultiplayerData.cs:128]: Backstory was turned into a `Def`, so I removed this.
